### PR TITLE
[boschshc] Support smoke detector

### PR DESCRIPTION
**Pull request by GerdZanker**
Enhances the Bosch Smart Home addon with the possible support for smoke detector devices.
You can only do a smoke detector check and get the battery status.

The existing Bosch SHC API does not support fire alarm or activate the alarm siren.

Existing code from @coeing was cherry-picked from [old BoschSHC AddOn development repository](https://github.com/stefan-kaestle/openhab2-addons) and updated to work with latest main branch.

### CODE REVIEW

**Comment by GerdZanker** [comment]
Implements #13759

**Comment by GerdZanker** [comment]
Hello @david-pace, I used your new battery state code together with an older pending extension from @coeing to support smoke detector devices. 
See also the [old issue and related PR](https://github.com/stefan-kaestle/openhab2-addons/issues/51) in the initial Bosch SHC repo.

Can you please join this pull request to check if I "merged" the old code in the right way?

**Comment by jlaur** [comment]
@GerdZanker - thanks for your contribution. Can I ask you to regenerate the i18n properties file? See https://www.openhab.org/docs/developer/utils/i18n.html#generating-i18n-properties-file

Besides this, it looks good to me. I'll wait for @david-pace's response before merging.

File: bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/devices/smokedetector/SmokedetectorHandlerTest.java
Comment (david-pace) [inline bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/devices/smokedetector/SmokedetectorHandlerTest.java:27]: You could CamelCase this class name that it aligns with the class under test: `SmokeDetectorHandlerTest`

File: bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/devices/smokedetector/SmokedetectorHandlerTest.java
Comment (david-pace) [inline bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/devices/smokedetector/SmokedetectorHandlerTest.java:43]: You could add tests that actually test the functionality that your handler provides (in this case the smoke check actions). Refer to `testHandlerCommand_*` and `testUpdateChannel_*` methods in `SmartBulbHandlerTest` for examples.

File: bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/smokedetector/SmokeDetectorHandler.java
Comment (david-pace) [inline bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/smokedetector/SmokeDetectorHandler.java:33]: Optional enhancement: you could make this class abstract and then let a new SmokeDetectorHandler extend it, and also let TwinguardHandler extend it.

**Comment by david-pace** [comment]
Hi @GerdZanker, the code looks very good. I made some comments for optional enhancements that would increase the test coverage and minimally reduce code duplications.

**Comment by jlaur** [comment]
@GerdZanker - did you see @david-pace's comments? Last milestone for 3.4 is near, but it's still possible to make it, if you would find time to have a look at the feedback.

**Comment by GerdZanker** [comment]
> @GerdZanker - did you see @david-pace's comments? Last milestone for 3.4 is near, but it's still possible to make it, if you would find time to have a look at the feedback.

Hello @jlaur, 
yes, I'm aware of all your pull request findings and I'm working on the suggestions.
But it takes time for me, because I left already my small Java comfort zone. E.g. I learn new stuff about writing tests.

**Comment by david-pace** [comment]
Thanks @GerdZanker, please let me know if you need any assistance with the tests. I could also add the tests later if it does not work out for some reason.

**Comment by GerdZanker** [comment]
> Hi @GerdZanker, the code looks very good. I made some comments for optional enhancements that would increase the test coverage and minimally reduce code duplications.

I hope my changes are the expected enhancements:
- i18n file is re-generated with maven
- used an enum based string to return the status and to request the smoke test
- abstract base classes are added
- new tests added, but copied and pasted test code lines from other tests without having any experience with Mockito
- renamed classes and filenames to stick to CamelCase pattern for SmokeDetector
  
> Thanks @GerdZanker, please let me know if you need any assistance with the tests. I could also add the tests later if it does not work out for some reason.

Yes @david-pace your review, improvements and even tests are of cause welcome.
If you find time, don't hesitate to change code and push new commits because I can continue earliest next week.

File: bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/smokedetector/SmokeDetectorHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/smokedetector/SmokeDetectorHandler.java:23]:
```suggestion
 * @author Gerd Zanker - AbstractSmokeDetectorHandler refactoring for reuse
```

File: bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/twinguard/TwinguardHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/twinguard/TwinguardHandler.java:39]:
```suggestion
 * @author Gerd Zanker - AbstractSmokeDetectorHandler refactoring for reuse
```

File: bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/AbstractSmokeDetectorHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/AbstractSmokeDetectorHandler.java:32]:
```suggestion
 * @author Gerd Zanker - AbstractSmokeDetectorHandler refactoring for reuse
```

**Review by jlaur — CHANGES_REQUESTED** [review]
This time I only found some minor typos, I hope @david-pace will be able to review latest commit in relation to resolved comments?

**Comment by david-pace** [comment]
Sure, I will review it tomorrow as soon as possible :+1:

File: bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/devices/AbstractSmokeDetectorHandlerTest.java
Comment (jlaur) [inline bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/devices/AbstractSmokeDetectorHandlerTest.java:44]: `@NonNullByDefault` annotation missing. This also yields a SAT warning.

File: bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/services/smokedetectorcheck/dto/SmokeDetectorCheckServiceState.java
Comment (jlaur) [inline bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/services/smokedetectorcheck/dto/SmokeDetectorCheckServiceState.java:16]: This import introduces a warning, because it's only used in Javadoc: `The import org.openhab.binding.boschshc.internal.services.smokedetectorcheck.SmokeDetectorCheckService is never used`

Perhaps the simplest fix would be to fully qualify it in Javadoc below and remove the import?

File: bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/devices/AbstractBatteryPoweredDeviceHandlerTest.java
Comment (david-pace) [reply bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/devices/AbstractBatteryPoweredDeviceHandlerTest.java:34]: I think it was removed due to the following compilation error:

```
Null constraint mismatch: The type 'T extends org.openhab.binding.boschshc.internal.devices.AbstractSmokeDetectorHandler' is not a valid substitute for the type parameter 'T extends org.openhab.binding.boschshc.internal.devices.@NonNull AbstractBatteryPoweredDeviceHandler'
```

But I think I found a way to fix this (will push a commit soon).

Comment (jlaur) [reply bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/devices/AbstractBatteryPoweredDeviceHandlerTest.java:34]: I checked, and I'm guessing it was related to this error: 
```
AbstractSmokeDetectorHandlerTest.java:[45,57] Null constraint mismatch: The type 'T extends org.openhab.binding.boschshc.internal.devices.AbstractSmokeDetectorHandler' is not a valid substitute for the type parameter 'T extends org.openhab.binding.boschshc.int`ernal.devices.@NonNull AbstractBatteryPoweredDeviceHandler
```

And then this when applying annotation in AbstractSmokeDetectorHandlerTest also: 
```
The @NonNull field smokeDetectorCheckStateCaptor may not have been initialized
```

I will provide a fix for that since I know you are currently away.

**Review by jlaur — APPROVED** [review]
Thanks for the new feature!

File: bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/devices/AbstractBatteryPoweredDeviceHandlerTest.java
Comment (david-pace) [reply bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/devices/AbstractBatteryPoweredDeviceHandlerTest.java]: ok, you were faster 😉

**Comment by jlaur** [comment]
> I've added two commits, so now incompetent to merge. Can one of you review those commits and merge the PR if everything looks good to you?

@wborn, @lolodomo perhaps?

**Comment by lolodomo** [comment]
> @openhab/add-ons-maintainers - I've added two commits, so now incompetent to merge. Can one of you review those commits and merge the PR if everything looks good to you?

Look only at your two last commits and LGTM.

**Comment by lolodomo** [comment]
```

Signed-off-by: Christian Oeing <christian.oeing@slashgames.org>
Signed-off-by: Gerd Zanker <gerd.zanker@web.de>
Signed-off-by: David Pace <dev@davidpace.de>
Signed-off-by: Jacob Laursen <jacob-github@vindvejr.dk>
Co-authored-by: Christian Oeing <christian.oeing@slashgames.org>
Co-authored-by: David Pace <dev@davidpace.de>
Co-authored-by: Jacob Laursen <jacob-github@vindvejr.dk>
```

Who should I keep as author and co-author ?

**Comment by jlaur** [comment]
> Who should I keep as author and co-author ?

It seems that it was based on work by Christian, so you can keep all three and strip me. Thanks all!

**Comment by GerdZanker** [comment]
Thank you @david-pace and @jlaur to finish the pull request. You improved it and taught my new java things I will use for future pull requests.

**Pull request merged**
